Add tag key in meta.yaml spec

ryanking13 opened the pull request:
### Description

I am preparing to make `pyodide build-recipes` CLI to work out-of-tree, so people can write and test recipes in their own repository. This PR is a first step to make it work.

This adds a new `tag` key in meta.yaml spec. A package can have multiples tags, and it is a metadata used to group packages.
For example, `PYODIDE_PACKAGES="tag-name" make` or `pyodide build-recipes tag-name` will build packages which contain that tag.

I removed the hard coded list of `ALWAYS_PACKAGES`, `CORE_PACKAGES`, `CORE_SCIPY_PACKAGES` from pyodide-build, and instead added a `always`, `core`, `min-scipy-stack` tag to corresponding packages.
(Note that we previously had a private key `_tag` used for documentation rendering.)

Partially related: #3341 

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [x] Add / update tests
- [x] Add new / update outdated documentation

rth commented:
> pyodide build-packages tag-name will build packages which contain that tag.

I find it a bit strange that the argument could be a tag or a package name. The reason we can put core in packages to build is that it's a meta-package so still something that relates to a package. Maybe it's just a matter of vocabulary in the documentation. The other alternative would be to build via,
```
pyodide build-packages --tag tag-name
```
but I agree that's probably not very useful.

Another thing, could you specify a list of valid tags in [the pydantic model](https://github.com/pyodide/pyodide/blob/cc7a734c498298bc59dd31f59eaf917ee7688acb/pyodide-build/pyodide_build/io.py#L12)  (via `typing.Literal`) so that if one makes a typo it would be detected?

ryanking13 commented:
> ```
> pyodide build-packages --tag tag-name
> ```
> 
> 
> but I agree that's probably not very useful.

This could work, but it is not very compatible with our current build system that use `PYODIDE_PACKAGES="..." make`. Of course we can introduce another env variable for tags, but I'm a bit reluctant to it.

> Another thing, could you specify a list of valid tags in [the pydantic model](https://github.com/pyodide/pyodide/blob/cc7a734c498298bc59dd31f59eaf917ee7688acb/pyodide-build/pyodide_build/io.py#L12) (via typing.Literal) so that if one makes a typo it would be detected?

Well, there is no invalid tags. I was thinking that people can add any tags they want.

rth commented on pyodide-build/pyodide_build/recipe.py:
```diff
@@ -0,0 +1,95 @@
+def load_recipes(
+    recipe_dir: Path,
+    names_or_tags: Iterable[str],
+    load_always: bool = True,
```
Maybe we could make  it `load_always_tag` to be less confusing

rth commented on docs/development/meta-yaml.md:
```diff
@@ -45,6 +45,16 @@ Some packages may have multiple top-level import names.
+The list of tag of the package. This is a meta information used to group
+the package. Normally this is not needed, but it can be useful to group
+packages by functionality. For example, the `min-scipy-stack` tag is used
+to group packages that are part of the minimal scipy stack.
```
It could we be useful to list what are currently used tags

rth commented on pyodide-build/pyodide_build/recipe.py:
```diff
@@ -0,0 +1,95 @@
+    -------
+    recipes
+        Dictionary of package name => config
+    flags
```
I don't fully understand why we need to additionally introduce the notion of flags. We could return the list of tags here, and check if `no-numpy-dependents` is in it, can't we? 

Or at least give it some more explicit name, because it's easy to confuse flags and tags, especially if one has a subset of keys of the other.

ryanking13 replied:
I wanted to hide details about tags / meta-packages from buildall, but I think it was a bad choice. I removed flags and let `generate_dependency_graph` deal with no-numpy-dependents.

rth reviewed (commented):
> Maybe we can add prefix for tags like, pyodide build-recipes pkg1 pkg2 tag:tag1 tag:tag2?

+1 for that.

> Well, there is no invalid tags. I was thinking that people can add any tags they want.

OK, but IMO we should still document currently used tags and link with the valid list in https://pyodide.org/en/stable/development/building-from-sources.html?highlight=min-scipy#partial-builds

Because a tag is currently only useful if `pyodide-build` has some internal logic that acts on the tag. Users can add arbitrary tags, but it will have no effect,  so what would be the use case there?  On the other hand, it's easy to mistype a tag, and then packages will not build as expected, which is harder to debug.  But yeah, we can leave the current behavior no strong objection.

ryanking13 commented:
Okay, I think I addressed most of the suggestions. Thanks for the detailed review @rth!

> Because a tag is currently only useful if pyodide-build has some internal logic that acts on the tag. Users can add arbitrary tags, but it will have no effect, so what would be the use case there? On the other hand, it's easy to mistype a tag, and then packages will not build as expected, which is harder to debug. But yeah, we can leave the current behavior no strong objection.

Yeah, I think most of the users will not use tags (unless someone makes a private set of packages) and this is basically for internal use. The primary goal for this PR is to reduce the dependency between pyodide-build and the pyodide main repository tree so that we are not bothered by ALWAYS and CORE packages when running pyodide-build out-of-tree.

hoodmane commented on pyodide-build/pyodide_build/recipe.py:
```diff
@@ -0,0 +1,101 @@
+            recipes[name_or_tag] = available_recipes[name_or_tag].copy(deep=True)
+
+        # 2. tag
+        elif name_or_tag.startswith("tag:") and name_or_tag[4:] in tagged_recipes:
+            tag = name_or_tag[4:]
```
```suggestion
        elif (
            name_or_tag.startswith("tag:")
            and (tag := name_or_tag.remove_prefix("tag:")) in tagged_recipes
        ):
```

hoodmane reviewed (approved):
Thanks @ryanking13! This is a nice change.

ryanking13 commented:
Thanks for the reviews!
